## You're here! 
Welcome to your first competition in the [ITI's AI Pro training program](https://ai.iti.gov.eg/epita/ai-engineer/)! We hope you enjoy and learn as much as we did prepairing this competition.


## Introduction

In the competition, it's required to predict the `Severity` of a car crash given info about the crash, e.g., location.

This is the getting started notebook. Things are kept simple so that it's easier to understand the steps and modify it.

Feel free to `Fork` this notebook and share it with your modifications **OR** use it to create your submissions.

### Prerequisites
You should know how to use python and a little bit of Machine Learning. You can apply the techniques you learned in the training program and submit the new solutions! 

### Checklist
You can participate in this competition the way you perefer. However, I recommend following these steps if this is your first time joining a competition on Kaggle.

* Fork this notebook and run the cells in order.
* Submit this solution.
* Make changes to the data processing step as you see fit.
* Submit the new solutions.

*You can submit up to 5 submissions per day. You can select only one of the submission you make to be considered in the final ranking.*


Don't hesitate to leave a comment or contact me if you have any question!

## Import the libraries

We'll use `pandas` to load and manipulate the data. Other libraries will be imported in the relevant sections.

In [1]:
import pandas as pd
import os

## Exploratory Data Analysis
In this step, one should load the data and analyze it. However, I'll load the data and do minimal analysis. You are encouraged to do thorough analysis!

Let's load the data using `pandas` and have a look at the generated `DataFrame`.

In [2]:
dataset_path = '/kaggle/input/car-crashes-severity-prediction/'

df = pd.read_csv(os.path.join(dataset_path, 'train.csv'))

print("The shape of the dataset is {}.\n\n".format(df.shape))

df.head()

The shape of the dataset is (6407, 16).




ID        Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
0   0  37.762150 -122.405660  False         0.044     False     False   
1   1  37.719157 -122.448254  False         0.000     False     False   
2   2  37.808498 -122.366852  False         0.000     False     False   
3   3  37.785930 -122.391080  False         0.009     False     False   
4   4  37.719141 -122.448457  False         0.000     False     False   

   Junction  No_Exit  Railway  Roundabout   Stop  Amenity Side  Severity  \
0     False    False    False       False  False     True    R         2   
1     False    False    False       False  False    False    R         2   
2     False    False    False       False   True    False    R         3   
3      True    False    False       False  False    False    R         1   
4     False    False    False       False  False    False    R         2   

             timestamp  
0  2016-03-25 15:13:02  
1  2020-05-05 19:23:00  
2  2016-09-16 19:57:16  
3  2020-03-29 19:48:43  
4  2019-10-09 08:47:00

In [3]:
df['Year'] = [d[0:4] for d in df['timestamp']]
df['Day']=[d[8:10] for d in df['timestamp']]
df['Month']=[d[5:7] for d in df['timestamp']]
df['Hour']=[d[10:13] for d in df['timestamp']]

df[['Day']] = df[['Day']].astype(int)
df[['Month']] = df[['Month']].astype(int)
df[['Hour']] = df[['Hour']].astype(int)
df[['Year']] = df[['Year']].astype(int)



df.head()

ID        Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
0   0  37.762150 -122.405660  False         0.044     False     False   
1   1  37.719157 -122.448254  False         0.000     False     False   
2   2  37.808498 -122.366852  False         0.000     False     False   
3   3  37.785930 -122.391080  False         0.009     False     False   
4   4  37.719141 -122.448457  False         0.000     False     False   

   Junction  No_Exit  Railway  Roundabout   Stop  Amenity Side  Severity  \
0     False    False    False       False  False     True    R         2   
1     False    False    False       False  False    False    R         2   
2     False    False    False       False   True    False    R         3   
3      True    False    False       False  False    False    R         1   
4     False    False    False       False  False    False    R         2   

             timestamp  Year  Day  Month  Hour  
0  2016-03-25 15:13:02  2016   25      3    15  
1  2020-05-05 19:23:00  2020    5      5    19  
2  2016-09-16 19:57:16  2016   16      9    19  
3  2020-03-29 19:48:43  2020   29      3    19  
4  2019-10-09 08:47:00  2019    9     10     8

In [4]:
df.shape

(6407, 20)

In [5]:
df=df.drop(columns=['timestamp'])


In [6]:
df_2 = pd.read_csv(os.path.join(dataset_path, 'weather-sfcsv.csv'))
df_2.head()


Year  Day  Month  Hour Weather_Condition  Wind_Chill(F)  Precipitation(in)  \
0  2020   27      7    18              Fair           64.0                0.0   
1  2017   30      9    17     Partly Cloudy            NaN                NaN   
2  2017   27      6     5          Overcast            NaN                NaN   
3  2016    7      9     9             Clear            NaN                NaN   
4  2019   19     10     2              Fair           52.0                0.0   

   Temperature(F)  Humidity(%)  Wind_Speed(mph)  Visibility(mi) Selected  
0            64.0         70.0             20.0            10.0       No  
1            71.1         57.0              9.2            10.0       No  
2            57.9         87.0             15.0             9.0       No  
3            66.9         73.0              4.6            10.0       No  
4            52.0         89.0              0.0             9.0       No

In [7]:
df_j = pd.merge(df,df_2,on=['Year','Day','Month','Hour'], how = 'outer')

In [8]:
df_j.shape

(9531, 27)

In [9]:
df_j.head()

ID       Lat        Lng   Bump  Distance(mi) Crossing Give_Way  \
0     0.0  37.76215 -122.40566  False         0.044    False    False   
1     0.0  37.76215 -122.40566  False         0.044    False    False   
2   803.0  37.76215 -122.40566  False         0.044    False    False   
3   803.0  37.76215 -122.40566  False         0.044    False    False   
4  4522.0  37.75232 -122.40303  False         0.306    False    False   

  Junction No_Exit Railway  ... Month Hour Weather_Condition Wind_Chill(F)  \
0    False   False   False  ...     3   15  Scattered Clouds           NaN   
1    False   False   False  ...     3   15     Partly Cloudy           NaN   
2    False   False   False  ...     3   15  Scattered Clouds           NaN   
3    False   False   False  ...     3   15     Partly Cloudy           NaN   
4    False   False   False  ...     3   15  Scattered Clouds           NaN   

   Precipitation(in)  Temperature(F)  Humidity(%)  Wind_Speed(mph)  \
0                NaN            64.0         58.0             23.0   
1                NaN            66.0         54.0             17.3   
2                NaN            64.0         58.0             23.0   
3                NaN            66.0         54.0             17.3   
4                NaN            64.0         58.0             23.0   

   Visibility(mi) Selected  
0            10.0       No  
1            10.0       No  
2            10.0       No  
3            10.0       No  
4            10.0       No  

[5 rows x 27 columns]

In [10]:
df_j.dropna(inplace=True)

We've got 6407 examples in the dataset with 14 featues, 1 ID, and the `Severity` of the crash.

By looking at the features and a sample from the data, the features look of numerical and catogerical types. What about some descriptive statistics?

In [11]:
df_j.drop(columns='ID').describe()
df_j[['Junction']] = df_j[['Junction']].astype(int)

stacked = df_j[['Weather_Condition']].stack()
df_j[['Weather_Condition']] = pd.Series(stacked.factorize()[0], index=stacked.index).unstack()

df_j["Weather_Condition"].fillna( (df_j['Weather_Condition'].mean()), inplace = True)


In [12]:
df_j.dropna(inplace=True)

In [13]:
df_j.shape

(4636, 27)

The output shows desciptive statistics for the numerical features, `Lat`, `Lng`, `Distance(mi)`, and `Severity`. I'll use the numerical features to demonstrate how to train the model and make submissions. **However you shouldn't use the numerical features only to make the final submission if you want to make it to the top of the leaderboard.**

## Data Splitting

Now it's time to split the dataset for the training step. Typically the dataset is split into 3 subsets, namely, the training, validation and test sets. In our case, the test set is already predefined. So we'll split the "training" set into training and validation sets with 0.8:0.2 ratio. 

*Note: a good way to generate reproducible results is to set the seed to the algorithms that depends on randomization. This is done with the argument `random_state` in the following command* 

In [14]:
from sklearn.model_selection import train_test_split

train_df, val_df = train_test_split(df_j, test_size=0.2, random_state=42) # Try adding `stratify` here


X_train = train_df.drop(columns=['ID', 'Severity'])
y_train = train_df['Severity']

X_val = val_df.drop(columns=['ID', 'Severity'])
y_val = val_df['Severity']


In [15]:
df_j.shape


(4636, 27)

As pointed out eariler, I'll use the numerical features to train the classifier. **However, you shouldn't use the numerical features only to make the final submission if you want to make it to the top of the leaderboard.** 

In [16]:
# This cell is used to select the numerical features. IT SHOULD BE REMOVED AS YOU DO YOUR WORK.
X_train = X_train[['Lat', 'Lng', 'Distance(mi)','Weather_Condition','Humidity(%)','Visibility(mi)']]
X_val = X_val[['Lat', 'Lng', 'Distance(mi)','Weather_Condition','Humidity(%)','Visibility(mi)']]

## Model Training

Let's train a model with the data! We'll train a Random Forest Classifier to demonstrate the process of making submissions. 

In [17]:
from sklearn.ensemble import RandomForestClassifier

# Create an instance of the classifier
classifier = RandomForestClassifier(max_depth=2, random_state=0)

# Train the classifier
classifier = classifier.fit(X_train, y_train)

Now let's test our classifier on the validation dataset and see the accuracy.

In [18]:
print("The accuracy of the classifier on the validation set is ", (classifier.score(X_val, y_val)))

The accuracy of the classifier on the validation set is  0.7672413793103449


Well. That's a good start, right? A classifier that predicts all examples' `Severity` as 2 will get around 0.63. You should get better score as you add more features and do better data preprocessing.

## Submission File Generation

We have built a model and we'd like to submit our predictions on the test set! In order to do that, we'll load the test set, predict the class and save the submission file. 

First, we'll load the data.

In [19]:
test_df = pd.read_csv(os.path.join(dataset_path, 'test.csv'))
test_df.head()

ID        Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
0  6407  37.786060 -122.390900  False         0.039     False     False   
1  6408  37.769609 -122.415057  False         0.202     False     False   
2  6409  37.807495 -122.476021  False         0.000     False     False   
3  6410  37.761818 -122.405869  False         0.000     False     False   
4  6411  37.732350 -122.414100  False         0.670     False     False   

   Junction  No_Exit  Railway  Roundabout   Stop  Amenity Side  \
0      True    False    False       False  False    False    R   
1     False    False    False       False  False    False    R   
2     False    False    False       False  False    False    R   
3      True    False    False       False  False    False    R   
4     False    False    False       False  False    False    R   

             timestamp  
0  2016-04-04 19:20:31  
1  2020-10-28 11:51:00  
2  2019-09-09 07:36:45  
3  2019-08-06 15:46:25  
4  2018-10-17 09:54:58

Note that the test set has the same features and doesn't have the `Severity` column.
At this stage one must **NOT** forget to apply the same processing done on the training set on the features of the test set.

Now we'll add `Severity` column to the test `DataFrame` and add the values of the predicted class to it.

**I'll select the numerical features here as I did in the training set. DO NOT forget to change this step as you change the preprocessing of the training data.**

In [20]:
test_df['Year'] = [d[0:4] for d in test_df['timestamp']]
test_df['Day']=[d[8:10] for d in test_df['timestamp']]
test_df['Month']=[d[5:7] for d in test_df['timestamp']]
test_df['Hour']=[d[10:13] for d in test_df['timestamp']]

test_df[['Day']] = test_df[['Day']].astype(int)
test_df[['Month']] = test_df[['Month']].astype(int)
test_df[['Hour']] = test_df[['Hour']].astype(int)
test_df[['Year']] = test_df[['Year']].astype(int)



test_df.head()

ID        Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
0  6407  37.786060 -122.390900  False         0.039     False     False   
1  6408  37.769609 -122.415057  False         0.202     False     False   
2  6409  37.807495 -122.476021  False         0.000     False     False   
3  6410  37.761818 -122.405869  False         0.000     False     False   
4  6411  37.732350 -122.414100  False         0.670     False     False   

   Junction  No_Exit  Railway  Roundabout   Stop  Amenity Side  \
0      True    False    False       False  False    False    R   
1     False    False    False       False  False    False    R   
2     False    False    False       False  False    False    R   
3      True    False    False       False  False    False    R   
4     False    False    False       False  False    False    R   

             timestamp  Year  Day  Month  Hour  
0  2016-04-04 19:20:31  2016    4      4    19  
1  2020-10-28 11:51:00  2020   28     10    11  
2  2019-09-09 07:36:45  2019    9      9     7  
3  2019-08-06 15:46:25  2019    6      8    15  
4  2018-10-17 09:54:58  2018   17     10     9

In [21]:
df_g = pd.merge(test_df,df_2,on=['Year','Day','Month','Hour'], how = 'outer')

In [22]:
df_g.head()

ID        Lat         Lng   Bump  Distance(mi) Crossing Give_Way  \
0  6407.0  37.786060 -122.390900  False         0.039    False    False   
1  6408.0  37.769609 -122.415057  False         0.202    False    False   
2  6693.0  37.769585 -122.414182  False         0.316    False    False   
3  7194.0  37.769669 -122.416109  False         0.214    False    False   
4  6409.0  37.807495 -122.476021  False         0.000    False    False   

  Junction No_Exit Railway  ... Month Hour Weather_Condition Wind_Chill(F)  \
0     True   False   False  ...     4   19     Partly Cloudy           NaN   
1    False   False   False  ...    10   11              Fair          65.0   
2    False   False   False  ...    10   11              Fair          65.0   
3    False   False   False  ...    10   11              Fair          65.0   
4    False   False   False  ...     9    7     Mostly Cloudy          58.0   

  Precipitation(in)  Temperature(F)  Humidity(%)  Wind_Speed(mph)  \
0               NaN            63.0         60.0             10.4   
1               0.0            65.0         56.0              5.0   
2               0.0            65.0         56.0              5.0   
3               0.0            65.0         56.0              5.0   
4               0.0            58.0         90.0             18.0   

   Visibility(mi) Selected  
0            10.0       No  
1             9.0       No  
2             9.0       No  
3             9.0       No  
4            10.0       No  

[5 rows x 27 columns]

In [23]:
df_g.dropna(inplace=True)

In [24]:
df_g.drop(columns='ID').describe()
df_g[['Junction']] = df_g[['Junction']].astype(int)

stacked = df_j[['Weather_Condition']].stack()
df_g[['Weather_Condition']] = pd.Series(stacked.factorize()[0], index=stacked.index).unstack()

df_g["Weather_Condition"].fillna( (df_g['Weather_Condition'].mean()), inplace = True)

In [25]:
df_g.dropna(inplace=True)

In [26]:
X_test = df_g.drop(columns=['ID'])

# You should update/remove the next line once you change the features used for training
X_test = X_test[['Lat', 'Lng', 'Distance(mi)','Weather_Condition','Humidity(%)','Visibility(mi)']]

y_test_predicted = classifier.predict(X_test)

df_g['Severity'] = y_test_predicted

test_df.head()

ID        Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
0  6407  37.786060 -122.390900  False         0.039     False     False   
1  6408  37.769609 -122.415057  False         0.202     False     False   
2  6409  37.807495 -122.476021  False         0.000     False     False   
3  6410  37.761818 -122.405869  False         0.000     False     False   
4  6411  37.732350 -122.414100  False         0.670     False     False   

   Junction  No_Exit  Railway  Roundabout   Stop  Amenity Side  \
0      True    False    False       False  False    False    R   
1     False    False    False       False  False    False    R   
2     False    False    False       False  False    False    R   
3      True    False    False       False  False    False    R   
4     False    False    False       False  False    False    R   

             timestamp  Year  Day  Month  Hour  
0  2016-04-04 19:20:31  2016    4      4    19  
1  2020-10-28 11:51:00  2020   28     10    11  
2  2019-09-09 07:36:45  2019    9      9     7  
3  2019-08-06 15:46:25  2019    6      8    15  
4  2018-10-17 09:54:58  2018   17     10     9

Now we're ready to generate the submission file. The submission file needs the columns `ID` and `Severity` only.

In [27]:
df_g[['ID', 'Severity']].to_csv('/kaggle/working/submission.csv', index=False)

The remaining steps is to submit the generated file and are as follows. 

1. Press `Save Version` on the upper right corner of this notebook.
2. Write a `Version Name` of your choice and choose `Save & Run All (Commit)` then click `Save`.
3. Wait for the saved notebook to finish running the go to the saved notebook.
4. Scroll down until you see the output files then select the `submission.csv` file and click `Submit`.

Now your submission will be evaluated and your score will be updated on the leaderboard! CONGRATULATIONS!!

## Conclusion

In this notebook, we have demonstrated the essential steps that one should do in order to get "slightly" familiar with the data and the submission process. We chose not to go into details in each step to keep the welcoming notebook simple and make a room for improvement.

You're encourged to `Fork` the notebook, edit it, add your insights and use it to create your submission.